File: examples/client-gen/tictactoe/types/game_state.py
from __future__ import annotations
import typing
from dataclasses import dataclass
from solders.pubkey import Pubkey
from anchorpy.borsh_extension import EnumForCodegen, BorshPubkey
import borsh_construct as borsh


class WonJSONValue(typing.TypedDict):
    winner: str


class WonValue(typing.TypedDict):
    winner: Pubkey


class ActiveJSON(typing.TypedDict):
    kind: typing.Literal["Active"]


class TieJSON(typing.TypedDict):
    kind: typing.Literal["Tie"]


class WonJSON(typing.TypedDict):
    value: WonJSONValue
    kind: typing.Literal["Won"]


@dataclass
class Active:
    discriminator: typing.ClassVar = 0
    kind: typing.ClassVar = "Active"

    @classmethod
    def to_json(cls) -> ActiveJSON:
        return ActiveJSON(
            kind="Active",
        )

    @classmethod
    def to_encodable(cls) -> dict:
        return {
            "Active": {},
        }


@dataclass
class Tie:
    discriminator: typing.ClassVar = 1
    kind: typing.ClassVar = "Tie"

    @classmethod
    def to_json(cls) -> TieJSON:
        return TieJSON(
            kind="Tie",
        )

    @classmethod
    def to_encodable(cls) -> dict:
        return {
            "Tie": {},
        }


@dataclass
class Won:
    discriminator: typing.ClassVar = 2
    kind: typing.ClassVar = "Won"
    value: WonValue

    def to_json(self) -> WonJSON:
        return WonJSON(
            kind="Won",
            value={
                "winner": str(self.value["winner"]),
            },
        )

    def to_encodable(self) -> dict:
        return {
            "Won": {
                "winner": self.value["winner"],
            },
        }


GameStateKind = typing.Union[Active, Tie, Won]
GameStateJSON = typing.Union[ActiveJSON, TieJSON, WonJSON]


def from_decoded(obj: dict) -> GameStateKind:
    if not isinstance(obj, dict):
        raise ValueError("Invalid enum object")
    if "Active" in obj:
        return Active()
    if "Tie" in obj:
        return Tie()
    if "Won" in obj:
        val = obj["Won"]
        return Won(
            WonValue(
                winner=val["winner"],
            )
        )
    raise ValueError("Invalid enum object")


def from_json(obj: GameStateJSON) -> GameStateKind:
    if obj["kind"] == "Active":
        return Active()
    if obj["kind"] == "Tie":
        return Tie()
    if obj["kind"] == "Won":
        won_json_value = typing.cast(WonJSONValue, obj["value"])
        return Won(
            WonValue(
                winner=Pubkey.from_string(won_json_value["winner"]),
            )
        )
    kind = obj["kind"]
    raise ValueError(f"Unrecognized enum kind: {kind}")


layout = EnumForCodegen(
    "Active" / borsh.CStruct(),
    "Tie" / borsh.CStruct(),
    "Won" / borsh.CStruct("winner" / BorshPubkey),
)

File: examples/tic-tac-toe/generated-client/types/GameState.ts
import { PublicKey } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh"

export interface ActiveJSON {
  kind: "Active"
}

export class Active {
  readonly discriminator = 0
  readonly kind = "Active"

  toJSON(): ActiveJSON {
    return {
      kind: "Active",
    }
  }

  toEncodable() {
    return {
      Active: {},
    }
  }
}

export interface TieJSON {
  kind: "Tie"
}

export class Tie {
  readonly discriminator = 1
  readonly kind = "Tie"

  toJSON(): TieJSON {
    return {
      kind: "Tie",
    }
  }

  toEncodable() {
    return {
      Tie: {},
    }
  }
}

export type WonFields = {
  winner: PublicKey
}
export type WonValue = {
  winner: PublicKey
}

export interface WonJSON {
  kind: "Won"
  value: {
    winner: string
  }
}

export class Won {
  readonly discriminator = 2
  readonly kind = "Won"
  readonly value: WonValue

  constructor(value: WonFields) {
    this.value = {
      winner: value.winner,
    }
  }

  toJSON(): WonJSON {
    return {
      kind: "Won",
      value: {
        winner: this.value.winner.toString(),
      },
    }
  }

  toEncodable() {
    return {
      Won: {
        winner: this.value.winner,
      },
    }
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function fromDecoded(obj: any): types.GameStateKind {
  if (typeof obj !== "object") {
    throw new Error("Invalid enum object")
  }

  if ("Active" in obj) {
    return new Active()
  }
  if ("Tie" in obj) {
    return new Tie()
  }
  if ("Won" in obj) {
    const val = obj["Won"]
    return new Won({
      winner: val["winner"],
    })
  }

  throw new Error("Invalid enum object")
}

export function fromJSON(obj: types.GameStateJSON): types.GameStateKind {
  switch (obj.kind) {
    case "Active": {
      return new Active()
    }
    case "Tie": {
      return new Tie()
    }
    case "Won": {
      return new Won({
        winner: new PublicKey(obj.value.winner),
      })
    }
  }
}

export function layout(property?: string) {
  const ret = borsh.rustEnum([
    borsh.struct([], "Active"),
    borsh.struct([], "Tie"),
    borsh.struct([borsh.publicKey("winner")], "Won"),
  ])
  if (property !== undefined) {
    return ret.replicate(property)
  }
  return ret
}
